pkg/labels/cidr/cidr_test.go

```diff
@@ -309,6 +320,124 @@ func BenchmarkLabels_SortedListCIDRIDs(b *testing.B) {
 	}
 }
 
+func BenchmarkGetCIDRLabelsConcurrent(b *testing.B) {
+	prefixes := make([]netip.Prefix, 0, 16)
+	octets := [4]byte{0, 0, 1, 1}
+	for i := 0; i < 16; i++ {
+		octets[0], octets[1] = byte(rand.Intn(256)), byte(rand.Intn(256))
+		prefixes = append(prefixes, netip.PrefixFrom(netip.AddrFrom4(octets), 32))
+	}
+
+	for _, goroutines := range []int{1, 2, 4, 16, 32, 48} {
+		b.Run(strconv.Itoa(goroutines), func(b *testing.B) {
+			b.ResetTimer()
+			for i := 0; i < b.N; i++ {
+				b.StopTimer()
+				start := make(chan struct{})
+				var wg sync.WaitGroup
+
+				wg.Add(goroutines)
+				for j := 0; j < goroutines; j++ {
+					go func() {
+						defer wg.Done()
+
+						<-start
+
+						for k := 0; k < 64; k++ {
+							_ = GetCIDRLabels(prefixes[rand.Intn(len(prefixes))])
+						}
+					}()
+				}
+
+				b.StartTimer()
+				close(start)
+				wg.Wait()
+			}
+		})
+	}
+}
+
+// BenchmarkCIDRLabelsCacheHeapUsageIPv4 should be run with -benchtime=1x
+func BenchmarkCIDRLabelsCacheHeapUsageIPv4(b *testing.B) {
+	b.Skip()
+
+	// clear the cache
+	cidrLabelsCache, _ = simplelru.NewLRU[netip.Prefix, []labels.Label](cidrLabelsCacheMaxSize, nil)
+
+	// be sure to fill the cache
+	prefixes := make([]netip.Prefix, 0, 256*256)
+	octets := [4]byte{0, 0, 1, 1}
+	for i := 0; i < 256*256; i++ {
+		octets[0], octets[1] = byte(i/256), byte(i%256)
+		prefixes = append(prefixes, netip.PrefixFrom(netip.AddrFrom4(octets), 32))
+	}
+
+	var m1, m2 runtime.MemStats
+	// One GC does not give precise results,
+	// because concurrent sweep may be still in progress.
+	runtime.GC()
+	runtime.GC()
+	runtime.ReadMemStats(&m1)
+
+	b.ResetTimer()
+
+	for i := 0; i < b.N; i++ {
+		for _, cidr := range prefixes {
+			_ = GetCIDRLabels(cidr)
+		}
+	}
+	b.StopTimer()
+
+	runtime.GC()
+	runtime.GC()
+	runtime.ReadMemStats(&m2)
+
+	usage := m2.HeapAlloc - m1.HeapAlloc
+	b.Logf("Memoization map heap usage: %.2f KiB", float64(usage)/1024)
```

Comment:
unless you were aware, you can also plumb this into the benchmark result to make it easier to use things like benchstat to compare it; https://github.com/cilium/cilium/pull/21895/files#diff-a5084c37ba77e5fe28cbc8d15ffe1e5c312e2daf79c8a363ec52df758818d26dR1091

Reply:
Thanks for the heads up, but I think that in this case it is better to switch to a simple test instead of a benchmark. That's because we want to measure the memory usage when the cache is full and to do that with the benchmark we would need to either:

- pass `benchtime=1x` to complete only a single pass
- reset the cache at each pass as part of the code to benchmark (but this resulted in longer benchmark run, probably for the time needed to stabilize the time measure)